### DISCUSSION TITLE
How to stream processing using WasmEdge ans WASI-NN?

### DISCUSSION THREAD

**Discussion by KeeganJin** [WasmEdge/WasmEdge#2771] [Q&A] [answered]
Hi all,

I am working on a project to perform machine learning on incoming data streams in WASM. Currently, I am using Rust and WasmEdge, and thinking of using WASI-NN and embed function, which means embedding the WASM function (including data loading, model loading, and inferencing).   However, in this case, the model has to be loaded every time the function is called. And the function will be called for each record of the data stream, which will cost a huge I/O overhead. 
I am looking for an approach, that can be used to achieve a loading model once but inferencing infinitly. 
Does anyone have an idea or project that can be referred to?
Thanks

**Comment by juntao** [comment]
The mediapipe-rs project have examples of how to load a model once and use it repeatedly in a session. 

Another part of the problem is that we need to move model from file into Wasm and then out to the WASI NN backend. For this problem, this feature will help you bypass the Wasm memory. 

https://github.com/WasmEdge/WasmEdge/pull/2742

If these do not work for you, please get in touch so that we can learn about your exact use cases. Thanks.

**Reply by KeeganJin** [reply]
Thanks, I will investigate that first.

**Reply by KeeganJin** [reply]
Thanks for your reply. I checked the mediapipe-rs project. But it kind of varies from my use case. 
My intended use case (what I am currently searching and working on) is to build a **lib.rs** for image classification which contains functions, such as 
`classifier_builder(), infer(), image_to_tensor(),` 
wasi-nn is currently being considered.
Then compile the **lib.rs** into WASM (image_classification.wasm). Therefore, I think **wasmedge_bindgen** is necessary in this case. 

Then outside of the .wasm file, in the **main** function, we can apply 
`vm = VmBuilder::register_module(image_classification.wasm)`
`classifier = vm.run_func("classifier_builder", model_requirements_params);`
`for image in images{
infer(image)
}
`
I checked the embed func, wasmedge_bindgen examples as well as other examples, 
but not quite sure if it is a feasible way for this image classification case with WasmEdge and wasmedge_bindgen. 
Thanks in advance. 
BR

**Reply by juntao** [reply, marked as answer]
Sure. It sounds like you will need the [WasmEdge Rust SDK](https://github.com/WasmEdge/wasmedge-rust-sdk).
